Logging

styts opened the pull request:
* [x] Log to console.log in scripts
* [x] Replace winston with bunyan in the main app src
* [x] Change function calls thoughtout the codebase to new logging levels (`verbose -> info`)
* [x] Log to files by default in development
* [x] Add --option for logging to stdout in development
* [x] Document how to use the new tool's filtering features
* [x] log `src` in development
---
Tail all the logs:

    npm run logs

Or with a filter applied (`this` is the the logged object):

    npm run logs -- -c 'this.sql_time > 0'

The output is pretty-printed for both db queries and react actions.

styts committed:
Remove winston dependency in scripts

Log with `console.log` instead

styts committed:
Replace winston with bunyan

inukshuk commented on scripts/log.js:
```diff
@@ -1,6 +1,6 @@
 'use strict'
 
-const winston = require('winston')
+/* eslint-disable no-console */
```
We can put this in `scripts/.eslintrc` (in fact it should already be there in `master` if you rebase)

styts committed:
Cleanup

styts committed:
Add dtrace-provider

Required by bunyan.

Needs to be listed as a dependency, so it gets compiled/linked against Electron
headers when doing a `npm rebuild`, otherwise errors about NODE_MODULE_VERSION
will show up.

styts committed:
Move colors to devDependencies

only used in `scripts`, not in `src`

styts committed:
Clearer logger code

The fallthrough forces one to think too hard

inukshuk commented on src/common/log.js:
```diff
@@ -2,6 +2,7 @@
 
 const { join } = require('path')
 const { assign } = Object
```
You can usually use spread properties everywhere in the Electron code instead of assign (everywhere but in `scripts`).

styts replied:
I tried to make use of `...` but could not make it work, as the module would export an object, not a function.

inukshuk replied:
Ah, yes, I didn't see how it was used, sorry -- in the rare case where you need to modify an object in place assign is good.

inukshuk commented on src/common/log.js:
```diff
@@ -2,6 +2,7 @@
 
 const { join } = require('path')
 const { assign } = Object
+const { existsSync, mkdirSync } = require('fs')
```
We will likely need to use `mkdirp` here then you don't need to use `exists` either, which is discouraged (deprecated even?) anyway. I.e., `const { sync: mdkir } = require('mkdirp')`

Since the logging runs at startup I'd prefer to use the async version, but since pretty much everything depends on the logger it's probably OK to make an exception.

inukshuk commented on src/common/log.js:
```diff
@@ -11,30 +12,39 @@ const logger = bunyan.createLogger({
+
+    logger.addStream({
+      type: 'rotating-file',
+      path: join(logDir, `${process.type}.log`),
```
Just a heads up that we'll have to make this configurable (we will have multiple 'renderer' processes using the logger).

inukshuk commented on src/common/log.js:
```diff
@@ -11,30 +12,39 @@ const logger = bunyan.createLogger({
-function init(dir) {
+function logToFolder(dir) {
+  if (dir) {
+    let logDir = join(dir, 'log')
```
Let's assume that the full directory path is passed-in. I think it makes it less flexible if we always add 'log' here.

styts committed:
Ensure both logging and top-level folder

styts committed:
Use mkdirp

styts committed:
Use all/only bunyan's logging levels

styts committed:
Change log.verbose calls to info

To be consistent with bunyan's target levels

styts committed:
Merge branch 'master' into logging

styts committed:
Logging improvements

styts committed:
Log db queries and time

styts committed:
Add `logs` command
